clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -598,65 +575,76 @@ public LeaveGroupResponse parse(ClientResponse response) {
-                    || error == Errors.NOT_COORDINATOR_FOR_GROUP) {
-                log.debug("Attempt to heart beat failed for group {} since coordinator {} is either not started or not valid.",
-                        groupId, coordinator);
-                coordinatorDead();
-                future.raise(error);
-            } else if (error == Errors.REBALANCE_IN_PROGRESS) {
-                log.debug("Attempt to heart beat failed for group {} since it is rebalancing.", groupId);
-                AbstractCoordinator.this.rejoinNeeded = true;
-                future.raise(Errors.REBALANCE_IN_PROGRESS);
-            } else if (error == Errors.ILLEGAL_GENERATION) {
-                log.debug("Attempt to heart beat failed for group {} since generation id is not legal.", groupId);
-                AbstractCoordinator.this.rejoinNeeded = true;
-                future.raise(Errors.ILLEGAL_GENERATION);
-            } else if (error == Errors.UNKNOWN_MEMBER_ID) {
-                log.debug("Attempt to heart beat failed for group {} since member id is not valid.", groupId);
-                memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
-                AbstractCoordinator.this.rejoinNeeded = true;
-                future.raise(Errors.UNKNOWN_MEMBER_ID);
-            } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
-                future.raise(new GroupAuthorizationException(groupId));
-            } else {
-                future.raise(new KafkaException("Unexpected error in heartbeat response: " + error.message()));
+            synchronized (AbstractCoordinator.this) {
+                sensors.heartbeatLatency.record(response.requestLatencyMs());
+                Errors error = Errors.forCode(heartbeatResponse.errorCode());
+                if (error == Errors.NONE) {
+                    log.debug("Received successful heartbeat response for group {}", groupId);
+                    future.complete(null);
+                } else if (error == Errors.GROUP_COORDINATOR_NOT_AVAILABLE
+                        || error == Errors.NOT_COORDINATOR_FOR_GROUP) {
+                    log.debug("Attempt to heart beat failed for group {} since coordinator {} is either not started or not valid.",
+                            groupId, coordinator);
+                    coordinatorDead();
+                    future.raise(error);
+                } else if (error == Errors.REBALANCE_IN_PROGRESS) {
+                    log.debug("Attempt to heart beat failed for group {} since it is rebalancing.", groupId);
+                    AbstractCoordinator.this.rejoinNeeded = true;
+                    future.raise(Errors.REBALANCE_IN_PROGRESS);
+                } else if (error == Errors.ILLEGAL_GENERATION) {
+                    log.debug("Attempt to heart beat failed for group {} since generation id is not legal.", groupId);
+                    AbstractCoordinator.this.rejoinNeeded = true;
+                    memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
```

Comment:
If we have an illegal generation, we've fallen out of the group, which means the coordinator doesn't know about our memberId anymore. When we send the next JoinGroup, the coordinator will just reject it with UNKNOWN_MEMBER_ID. So this saves a round trip for that case.